src/create-dashboard.js

```diff
@@ -63,34 +72,36 @@ module.exports = function (selector, diJSON, visOpts) {
 
   // TODO: We can probably move this logic somewhere else
   var widgetModels = [];
-  diJSON.widgets.forEach(function (widget) {
-    var widgetAttrs = _.omit(widget, 'dataview');
+  vizJSON.widgets.forEach(function (rawWidgetData) {
+    var layerId = rawWidgetData.layerId;
+    var widgetAttrs = _.omit(rawWidgetData, 'options', 'layerId');
+    var dataviewModel;
 
     // Create dataview if there's a definition provided
-    var dataviewModel;
-    var d = widget.dataview;
-    if (d) {
+    var dataviewAttrs = rawWidgetData.options;
+    if (dataviewAttrs) {
+      // TODO not ideal, should have a more maintainable way of mapping
+      dataviewAttrs.type = rawWidgetData.type === 'time-series'
+        ? 'histogram' // Time-series widget is represented by a histogram, so re-map the type
+        : rawWidgetData.type;
+
       var layer;
 
       // Find the Layer that the Widget should be created for.
-      // a layerId has top-priority, otherwise it tries with a layerIndex, and even a subLayerIndex (if available)
-      if (d.layerId) {
-        layer = vis.interactiveLayers.find(function (m) {
-          return d.layerId === m.get('id');
+      if (layerId) {
+        layer = vis.map.layers.find(function (m) {
+          return layerId === m.get('id');
         });
-      } else if (Number.isInteger(d.layerIndex)) {
-        layer = vis.map.layers.at(d.layerIndex);
-        if (layer && Number.isInteger(d.subLayerIndex)) {
-          layer = layer.layers.at(d.subLayerIndex);
-        }
+      } else if (Number.isInteger(rawWidgetData.layerIndex)) {
+        // TODO Since namedmap doesn't have ids we need to map in another way, here using index
+        //   should we solve this in another way?
+        layer = vis.map.layers.at(rawWidgetData.layerIndex);
```

Comment:
Have in mind that layer id is a concept that does not exist at Maps API level. Maps API is agnostic in that sense because it does not need anything like that internally. That means that you can use an id attribute for the layers and Maps API will not intervene modifying/removing it.

Also: we most likely will implement it as an optional attribute in the near future: https://github.com/CartoDB/Windshaft/issues/409. In that case we might need to impose some kind of validation as not all ids will be valid.

But the key thing here is that existing named maps don't include an id attribute either. So I would say it's that easy to just remove the int index option.